Philips.CodeAnalysis.MaintainabilityAnalyzers/Readability/AvoidMultipleLambdasOnSingleLineAnalyzer.cs

```diff
@@ -0,0 +1,87 @@
+﻿// © 2023 Koninklijke Philips N.V. See License.md in the project root for license information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Philips.CodeAnalysis.Common;
+
+namespace Philips.CodeAnalysis.MaintainabilityAnalyzers.Readability
+{
+	[DiagnosticAnalyzer(LanguageNames.CSharp)]
+	public class AvoidMultipleLambdasOnSingleLineAnalyzer : DiagnosticAnalyzer
+	{
+
+		private const string Title = @"Avoid putting multiple Lambda expressions on the same line";
+		public const string MessageFormat = @"Avoid putting multiple Lambda expressions on the same line";
+		private const string Description = @"Avoid putting multiple Lambda expressions on the same line. To improve readability, split them into separate lines.";
+		private const string Category = Categories.Readability;
+
+		private static readonly DiagnosticDescriptor Rule = new(Helper.ToDiagnosticId(DiagnosticIds.AvoidMultipleLambdasOnSingleLine), Title,
+			MessageFormat, Category, DiagnosticSeverity.Error, isEnabledByDefault: true, description: Description);
+
+
+		public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+		private readonly HashSet<string> visitedTypes = new();
+
+		public override void Initialize(AnalysisContext context)
+		{
+			context.EnableConcurrentExecution();
+			context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+			
+			context.RegisterSyntaxNodeAction(Analyze, SyntaxKind.SimpleLambdaExpression);
+			context.RegisterSyntaxNodeAction(Analyze, SyntaxKind.ParenthesizedLambdaExpression);
+			context.RegisterCompilationAction((ctx) => visitedTypes.Clear());
+		}
+
+		private void Analyze(SyntaxNodeAnalysisContext context)
+		{
+			SyntaxNode node = context.Node;
+			TypeDeclarationSyntax typeDeclaration =
+				node.Ancestors().OfType<TypeDeclarationSyntax>().FirstOrDefault();
+			
+			// Prevent reporting the same class twice.
```

Comment:
It seems OK to have multiple lambda findings per class. I would think we just want to avoid reporting for the same statement.

Reply:
We only report once now and the unit tests are sensitive to multiple diagnostics on a single statement.